### PR TITLE
HDPI-5806: enable counterclaim fee payment for defendant response (BE)

### PR DESCRIPTION
**Pull request by andrewdami11**
### Jira link

See [HDPI-5806](https://tools.hmcts.net/jira/browse/HDPI-5806)

Related FE PR: [pcs-frontend #1432](https://github.com/hmcts/pcs-frontend/pull/1432)

### Change description

Adds backend support for defendant counterclaim fee payment in respond-to-claim, so the CUI can create a service request at submit time and complete payment through GovPay.

When a defendant submits a response with a counterclaim and no Help with Fees reference:

- **Submit** creates a payment service request, resolves the fee from the fee register (amount-based counterclaim fee, ranged fee, or FEE0450 flat fee for “Something else”), and returns `serviceRequestReference` and `feeAmount` in the submit confirmation body.
- **Counterclaim persistence** saves the counterclaim with status `PENDING_COUNTER_CLAIM_ISSUED`.
- **Payment APIs** expose card-payment creation against a service request and status lookup by internal payment reference for the frontend return route.
- **Payment callback** routes counterclaim payments through `CounterClaimPaymentCallbackHandler`; on successful payment, updates counterclaim status to `COUNTER_CLAIM_ISSUED` and sets claim issued date.

Main development areas:

- **Submit flow** — `SubmitEventHandler` counterclaim payment branch, `RespondPossessionClaimSubmitResponse` confirmation payload, fee-type resolution (`COUNTER_CLAIM`, `COUNTER_CLAIM_RANGED`, `COUNTER_CLAIM_FLAT_FEE_FEE0450`).
- **Fees and pay** — `CounterClaimPaymentCallbackHandler`, `PaymentCallbackHandlerType.COUNTER_CLAIM_ISSUE`, `FeesAndPayTaskData.relatedEntityId`, and `PaymentService` updates for service-request lookup.
- **Schema** — Flyway `V114` renames `fee_payment.request_reference` to `service_request_reference` and adds `related_entity_id`.
- **Testing support cleanup** — removes `POST /testing-support/counterclaim/{counterclaimId}/status` and `POST /testing-support/defendant-response/{defendantResponseId}/status` as noted on the ticket.

### Testing done

Automated tests added or updated for:

- `SubmitEventHandlerTest` — service request creation, confirmation body, fee-type branches (including FEE0450 and ranged fees), HWF and non-counterclaim paths
- `CounterClaimPaymentCallbackHandlerTest` — paid and unsuccessful callback handling, status transitions
- `DefendantResponseServiceTest` — counterclaim saved as `PENDING_COUNTER_CLAIM_ISSUED`
- `PaymentControllerTest`, `PaymentServiceTest`, `PaymentCallbackStrategyFactoryTest`
- `TestingSupportControllerTest` — removed status endpoints
- `RespondPossessionClaimTest`, `PaymentCallBackControllerIT`

Manual GovPay / preview end-to-end verification is planned separately and is not evidenced in this PR body.

### Security Vulnerability Assessment

**CVE Suppression:** Are there any CVEs present in the codebase (either newly introduced or pre-existing) that are being intentionally suppressed or ignored by this commit?

- [ ] Yes
- [ ] No

### Checklist

- [x] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)
- [ ] Does this PR introduce a breaking change

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/service/CounterClaimPaymentCallbackHandler.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                null
+            );
+        }
+
+        CounterClaimEntity counterClaimEntity = counterClaimRepository.findById(counterClaimId)
+            .orElseThrow(() -> new IllegalArgumentException("Counterclaim not found: " + counterClaimId));
+
+        feePaymentEntity.setParty(counterClaimEntity.getParty());
+
+        if (PaymentStatus.PAID == feePaymentEntity.getPaymentStatus()) {
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/service/CounterClaimPaymentCallbackHandler.java:57]: I think this should also check that the counterclaim is not already in the COUNTER_CLAIM_ISSUED state. In fact, better to check that it is specifically in the state before that, (pending issue or whatever it is called), so that the state doesn't go backwards if a rogue callback came in sometime after the counterclaim had progressed to another state

Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/service/CounterClaimPaymentCallbackHandler.java:57]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/commits/a90f6dc453dec4fcc212ad4979ebbf0bd88edf69)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/domain/respondpossessionclaim/RespondPossessionClaimSubmitResponse.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.Data;
+
+import java.math.BigDecimal;
+
+@Builder
+@Data
+@AllArgsConstructor
+public class RespondPossessionClaimSubmitResponse {
+
+    private final CounterClaimStatus state;
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/domain/respondpossessionclaim/RespondPossessionClaimSubmitResponse.java]: Minor - this type would be better renamed to `CounterClaimState` for consistency

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java [hunk 3]
```diff
@@ -77,8 +106,114 @@ private SubmitResponse<State> validate(PossessionClaimResponse possessionClaimRe
...
-    private SubmitResponse<State> success() {
-        return SubmitResponse.defaultResponse();
+    private SubmitResponse<State> success(long caseReference, PossessionClaimResponse possessionClaimResponse) {
+        if (!isCounterClaimPaymentRequired(possessionClaimResponse)) {
+            return SubmitResponse.defaultResponse();
+        }
+
+        CounterClaim counterClaim = possessionClaimResponse.getDefendantResponses().getCounterClaim();
+        FeeType feeType = resolveCounterClaimFeeType(counterClaim);
+        FeeDetails feeDetails = feeService.getFee(feeType);
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: I think it would be cleaner to extract the fee calculation logic out of this class and into its own service. It would also make it easier to test the various scenarios, particularly around the boundary conditions for the fee codes.

Have a look at https://github.com/hmcts/pcs-api/blob/HDPI-4276_cui_payments_part2/src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/genapp/GenAppFeeCalculator.java for an example of a possible approach.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java [hunk 4]
```diff
@@ -77,8 +106,114 @@ private SubmitResponse<State> validate(PossessionClaimResponse possessionClaimRe
         return null;
     }
 
-    private SubmitResponse<State> success() {
-        return SubmitResponse.defaultResponse();
+    private SubmitResponse<State> success(long caseReference, PossessionClaimResponse possessionClaimResponse) {
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: This method name doesn't really describe what the method is doing.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java [hunk 5]
```diff
@@ -77,8 +106,114 @@ private SubmitResponse<State> validate(PossessionClaimResponse possessionClaimRe
...
+        if (!isCounterClaimPaymentRequired(possessionClaimResponse)) {
+            return SubmitResponse.defaultResponse();
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Don't you need to return a response body here as well to indicate to the frontend that the counter claim has been issued? Also is the initial counterclaim state actually being persisted with the entity?

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java [hunk 6]
```diff
@@ -77,8 +106,114 @@ private SubmitResponse<State> validate(PossessionClaimResponse possessionClaimRe
...
+            .paymentCallbackHandlerType(COUNTER_CLAIM_ISSUE)
+            .relatedEntityId(counterClaimEntityId)
+            .build();
+
+        PaymentServiceResponse paymentServiceResponse = paymentService.createServiceRequest(taskData);
+
+        RespondPossessionClaimSubmitResponse response = RespondPossessionClaimSubmitResponse.builder()
+            .state(CounterClaimStatus.PENDING_COUNTER_CLAIM_ISSUED)
+            .serviceRequestReference(paymentServiceResponse.getServiceRequestReference())
+            .feeAmount(feeDetails.getFeeAmount())
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: I wonder if this needs a slightly different model. The event is to respond to a possession claim - am I right in thinking that the possession claim is still "complete" even if there is a counterclaim element pending?

Perhaps this model just needs a CounterClaimSubmitResponse field within the RespondPossessionClaimSubmitResponse, to wrap the counterclaim specific response values

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java [hunk 7]
```diff
@@ -77,8 +106,114 @@ private SubmitResponse<State> validate(PossessionClaimResponse possessionClaimRe
...
+        UUID currentUserId = securityContextService.getCurrentUserId();
+        if (currentUserId == null) {
+            throw new IllegalStateException("Current user IDAM ID is null");
+        }
+        return partyService.getPartyEntityByIdamId(currentUserId, caseReference);
+    }
+
+    private UUID getPendingCounterClaimEntityId(long caseReference, UUID partyId) {
+        CounterClaimEntity counterClaimEntity =
+            counterClaimRepository.findFirstByPcsCaseCaseReferenceAndPartyIdAndStatusOrderByClaimSubmittedDateDesc(
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: This seems like quite a complicated, (and potentially unreliable), way to fetch the counterclaim that was raised at the same time as this response to claim.

Could you change the code above that calls `defendantResponseService.saveDefendantResponse(caseReference, responseDraftData)` so that it doesn't also create the counterclaim, (it is a distinct entity that can exist without a defendant response after all). 

Then extract that `saveCounterClaim` method code from `DefendantResponseService` into a new service that creates a counterclaim from the draft data, and return the exact CounterClaimEntity that was created.

File: src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/model/FeeType.java [hunk 8]
```diff
@@ -9,7 +9,10 @@ public enum FeeType {
     ENFORCEMENT_WARRANT_FEE("enforcementWarrantFee"),
     ENFORCEMENT_WRIT_FEE("enforcementWritFee"),
     GEN_APP_STANDARD_FEE("genAppStandardFee"),
-    GEN_APP_MAX_FEE("genAppMaxFee");
+    GEN_APP_MAX_FEE("genAppMaxFee"),
+    COUNTER_CLAIM_FLAT_FEE_FEE0450("counterClaimFlatFeeFEE0450"),
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/model/FeeType.java]: Minor - why does this name have the fee code in it unlike the others? In theory the fee code for a counterclaim flat fee might change in future, which is perhaps why we query with several parameters rather than the fee code directly.

File: src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/service/CounterClaimPaymentCallbackHandler.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            throw new PaymentCallbackException(
+                "Counterclaim payment callback missing relatedEntityId in task data",
+                null
+            );
+        }
+
+        CounterClaimEntity counterClaimEntity = counterClaimRepository.findById(counterClaimId)
+            .orElseThrow(() -> new IllegalArgumentException("Counterclaim not found: " + counterClaimId));
+
+        feePaymentEntity.setParty(counterClaimEntity.getParty());
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/service/CounterClaimPaymentCallbackHandler.java]: Minor - do you actually need to do this? I've been looking at this Party on the fee payment to see what it is used for and I suspect it isn't really needed. In any case, the fee payer isn't necessarily the party for a counterclaim, as a legal rep might have raised it on their behalf, (in a journey that is still to be implemented).

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/domain/respondpossessionclaim/RespondPossessionClaimSubmitResponse.java (hunk 2 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/domain/respondpossessionclaim/RespondPossessionClaimSubmitResponse.java]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/dd473f876b96977b9b953d2ff6ffeb62250033cf)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java (hunk 3 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/1376d53cc0c394a665a7fe308a4ce33318b032d2)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java (hunk 4 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/f436654f3ec3cb3d1f9a4a3b35dc02db2e9e4010)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java (hunk 5 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/f436654f3ec3cb3d1f9a4a3b35dc02db2e9e4010)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java (hunk 6 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/f436654f3ec3cb3d1f9a4a3b35dc02db2e9e4010)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java (hunk 7 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/f436654f3ec3cb3d1f9a4a3b35dc02db2e9e4010)

File: src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/service/CounterClaimPaymentCallbackHandler.java (hunk 9 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/service/CounterClaimPaymentCallbackHandler.java]: Removed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/4eb80febad70030e1d49242647108e02dbc719f8)

File: src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/model/FeeType.java (hunk 8 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/feesandpay/model/FeeType.java]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/1376d53cc0c394a665a7fe308a4ce33318b032d2)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java [hunk 10]
```diff
@@ -69,16 +102,86 @@ private SubmitResponse<State> validate(PossessionClaimResponse possessionClaimRe
...
     }
 
-    private SubmitResponse<State> success() {
-        return SubmitResponse.defaultResponse();
+    private SubmitResponse<State> buildSubmitResponse(
+        long caseReference,
+        PossessionClaimResponse possessionClaimResponse,
+        CounterClaimEntity counterClaimEntity
+    ) {
+        if (!isCounterClaimPaymentRequired(possessionClaimResponse)) {
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Is this `isCounterClaimPaymentRequired` method checking more than it needs to now? We know that if `counterClaimEntity` is not null then a counter claim was created, so the only check that is needed is whether HwF applies isn't it? (Rather than checking whether the defendant response included details of a counterclaim).

Also, (this is up to you 🙂 ), but personally I think it is cleaner to have that HwF logic and the fee lookup in the fee calculator class. Then this class just asks that calculator "what is the (optional) fee for this counter claim", and if there isn't one then you know that payment isn't required. That way, if/when the HwF only covers part of the fee, that can be encapsulated into the fee calculator and all this class needs to know is whether there is a fee and how much needs to be paid.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java (hunk 5 above)
Comment (scottstewart-sl) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: This is still returning the defaultResponse which would have no confirmation body. If there was a counterclaim with no payment due then I would expect the counterclaim entity state to be updated to Issued straight away, and that state returned in the confirmation body so that the frontend can see that no payment is due.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimFeeCalculator.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        }
+        if (counterClaim.getClaimType() == CounterClaimType.SOMETHING_ELSE) {
+            return FeeType.COUNTER_CLAIM_FLAT_FEE;
+        }
+
+        BigDecimal claimAmount = getCounterClaimAmount(counterClaim);
+        if (claimAmount == null || claimAmount.signum() < 0) {
+            return FeeType.COUNTER_CLAIM;
+        }
+        if (claimAmount.compareTo(new BigDecimal("5000")) <= 0) {
```
Comment (scottstewart-sl) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimFeeCalculator.java]: Minor - you could make this `new BigDecimal("5000")` into a constant with a suitable name.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java [hunk 12]
```diff
@@ -69,16 +102,86 @@ private SubmitResponse<State> validate(PossessionClaimResponse possessionClaimRe
...
+        FeesAndPayTaskData taskData = FeesAndPayTaskData.builder()
+            .feeDetails(feeDetails)
+            .ccdCaseNumber(String.valueOf(caseReference))
+            .caseReference(caseReference)
+            .responsiblePartyId(responsibleParty.getId())
+            .paymentCallbackHandlerType(COUNTER_CLAIM_ISSUE)
+            .relatedEntityId(counterClaimEntity.getId())
+            .build();
+
+        PaymentServiceResponse paymentServiceResponse = paymentService.createServiceRequest(taskData);
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: You're inside a DB transaction here, you should not be doing http requests.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java (hunk 10 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Refactored [here](https://github.com/hmcts/pcs-api/pull/1915/changes/a1e87af99c6a9f9793a0ba8e0b4ad0d5df136046)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimFeeCalculator.java (hunk 11 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimFeeCalculator.java]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/a1e87af99c6a9f9793a0ba8e0b4ad0d5df136046)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java (hunk 12 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/changes/8f9e27168c56233dae0de809ec4ea4633f0a9028)

Comment (scottstewart-sl) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/event/respondpossessionclaim/SubmitEventHandler.java]: Actually @linusnorton - there is already a DB transaction active for the event submit handler, added by `CaseSubmissionService` in the tooling. In this case we need to get the created service request reference to return to the caller, so we can't easily make this asynchronous either.

**Review by scottstewart-sl — APPROVED** [review]

**Review by linusnorton — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimFeeCalculator.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    private BigDecimal getCounterClaimAmountInPence(CounterClaim counterClaim) {
+        if (counterClaim.getIsClaimAmountKnown() == VerticalYesNo.YES) {
+            return counterClaim.getClaimAmount();
+        }
+        if (counterClaim.getIsClaimAmountKnown() == VerticalYesNo.NO) {
+            return counterClaim.getEstimatedMaxClaimAmount();
+        }
+        return null;
```
Comment (guygrewal77) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimFeeCalculator.java:53]: No coverage for this scenario

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimSubmitConfirmationService.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            throw new IllegalStateException("Current user IDAM ID is null");
+        }
+        return partyService.getPartyEntityByIdamId(currentUserId, caseReference);
+    }
+
+    private String writeAsString(RespondPossessionClaimSubmitResponse submitResponse) {
+        try {
+            return objectMapper.writeValueAsString(submitResponse);
+        } catch (JsonProcessingException e) {
+            throw new IllegalStateException("Unable to serialise respond possession claim submit response", e);
```
Comment (guygrewal77) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimSubmitConfirmationService.java:139]: This exception isn't currently covered in the tests.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/DefendantResponseService.java [hunk 15]
```diff
@@ -55,7 +48,6 @@ public class DefendantResponseService {
     private final PaymentAgreementService paymentAgreementService;
     private final DocumentService documentService;
     private final PartyAttributeAssertationService partyAttributeAssertationService;
-    private final CounterClaimRepository counterClaimRepository;
     private final Clock utcClock;
 
     public DefendantResponseService(PartyService partyService,
```
Comment (guygrewal77) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/DefendantResponseService.java:65]: There isn't any coverage for saveDefendantResponse() method.

File: src/main/resources/db/migration/V118__add_related_entity_id_to_fee_payment.sql [hunk 16]
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE fee_payment
+  RENAME COLUMN request_reference to service_request_reference;
```
Comment (guygrewal77) [inline src/main/resources/db/migration/V118__add_related_entity_id_to_fee_payment.sql:2]: There's a missing migration script number 114.

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/repository/CounterClaimRepository.java [hunk 17]
```diff
@@ -2,11 +2,23 @@
...
+import uk.gov.hmcts.reform.pcs.ccd.domain.respondpossessionclaim.CounterClaimState;
 import uk.gov.hmcts.reform.pcs.ccd.entity.respondpossessionclaim.CounterClaimEntity;
 
+import java.util.Optional;
 import java.util.UUID;
 
 @Repository
 public interface CounterClaimRepository extends JpaRepository<CounterClaimEntity, UUID> {
 
+    Optional<CounterClaimEntity> findFirstByPcsCaseCaseReferenceAndPartyIdAndStatusOrderByClaimSubmittedDateDesc(
```
Comment (guygrewal77) [inline src/main/java/uk/gov/hmcts/reform/pcs/ccd/repository/CounterClaimRepository.java:14]: These two methods are never accessed anywhere.

File: src/main/resources/db/migration/V118__add_related_entity_id_to_fee_payment.sql (hunk 16 above)
Comment (andrewdami11) [reply src/main/resources/db/migration/V118__add_related_entity_id_to_fee_payment.sql:2]: Addressed [here](https://github.com/hmcts/pcs-api/commit/25b64edbb)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimFeeCalculator.java (hunk 13 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimFeeCalculator.java:53]: Addressed [here](https://github.com/hmcts/pcs-api/commit/f499c0827)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimSubmitConfirmationService.java (hunk 14 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/CounterClaimSubmitConfirmationService.java:139]: Addressed [here](https://github.com/hmcts/pcs-api/commit/e04f36d15)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/DefendantResponseService.java (hunk 15 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/service/respondpossessionclaim/DefendantResponseService.java:65]: Addressed [here](https://github.com/hmcts/pcs-api/commit/0bbf7765b)

File: src/main/java/uk/gov/hmcts/reform/pcs/ccd/repository/CounterClaimRepository.java (hunk 17 above)
Comment (andrewdami11) [reply src/main/java/uk/gov/hmcts/reform/pcs/ccd/repository/CounterClaimRepository.java:14]: Addressed [here](https://github.com/hmcts/pcs-api/pull/1915/commits/1d1c3b653652a97271ec9c8eef0bd5f991a27cfa)

**Review by guygrewal77 — APPROVED** [review]
